haystack/preview/components/samplers/top_p.py

```diff
@@ -0,0 +1,202 @@
+    output = sampler.run(query=query, documents=docs)
+    docs = output["documents"]
+    assert len(docs) == 1
+    assert docs[0].text == "Berlin"
+    ```
+    """
+
+    def __init__(
+        self,
+        model_name_or_path: Optional[Union[str, Path]] = None,
+        top_p: Optional[float] = 1.0,
+        score_field: Optional[str] = "similarity_score",
+        device: Optional[str] = "cpu",
+    ):
+        """
+        Creates an instance of TopPSampler.
+
+        :param model_name_or_path: Path to a pre-trained sentence-transformers model. If not specified, no document
+        scoring will be performed, and it is assumed that documents already have scores in the metadata specified under
+        the `score_field` key.
+        :param top_p: Cumulative probability threshold for filtering the documents (usually between 0.9 and 0.99).
+        :param score_field: The name of the field that should be used to store the scores in a document's metadata.
+        :param device: torch device (for example, cuda:0, cpu, mps) to limit model inference to a specific device.
+        """
+        torch_and_transformers_import.check()
+        super().__init__()
+
+        self.model_name_or_path = model_name_or_path
+        self.top_p = top_p
+        self.score_field = score_field
+        self.device = device
+        self.model = None
+        self.tokenizer = None
+
+    def warm_up(self):
+        """
+        Warm up the model and tokenizer used in scoring the documents.
+        """
+        if self.model_name_or_path and not self._has_scoring_model():
+            self.model = AutoModelForSequenceClassification.from_pretrained(self.model_name_or_path)
+            self.model = self.model.to(self.device)
+            self.model.eval()
+            self.tokenizer = AutoTokenizer.from_pretrained(self.model_name_or_path)
+
+    def to_dict(self) -> Dict[str, Any]:
+        """
+        Serialize this component to a dictionary.
+        """
+        return default_to_dict(
+            self,
+            top_p=self.top_p,
+            score_field=self.score_field,
+            device=self.device,
+            model_name_or_path=self.model_name_or_path,
+        )
+
+    @classmethod
+    def from_dict(cls, data: Dict[str, Any]) -> "TopPSampler":
+        """
+        Deserialize this component from a dictionary.
+        """
+        return default_from_dict(cls, data)
+
+    @component.output_types(documents=List[Document])
+    def run(self, query: str, documents: List[Document], top_p: Optional[float] = None):
+        """
+        Returns a list of documents filtered using `top_p`, based on the similarity scores between the query and the
+        documents whose cumulative probability is less than or equal to `top_p`.
+
+        :param query: Query string.
+        :param documents: List of Documents.
+        :param top_p: Cumulative probability threshold for filtering the documents. If not provided, the top_p value
```

Comment:
I would add to this docstring if that top_p is not provided and if top_p is not provided at init then this will default to 1.0